Sources/FoundationEssentials/Data/Data+Reading.swift

```diff
@@ -0,0 +1,345 @@
+    internal static func _readBytesAndEncoding(fromPath path: String, maxLength: Int, encoding outEncoding: UnsafeMutablePointer<UInt>, bytes: UnsafeMutablePointer<UnsafeMutableRawPointer?>, length: UnsafeMutablePointer<Int>, didMap: UnsafeMutablePointer<ObjCBool>, options: Data.ReadingOptions, reportProgress: Bool) throws {
+        
+        var attrs: [String : Data] = [:]
+        let result = try readBytesFromFile(path: .path(path), reportProgress: reportProgress, maxLength: maxLength == Int.max ? nil : maxLength, options: options, attributesToRead: [NSFileAttributeStringEncoding], attributes: &attrs)
+        if let encodingAttributeData = attrs[NSFileAttributeStringEncoding] {
+            outEncoding.pointee = _NSEncodingFromDataForExtendedAttribute(encodingAttributeData)
+        } else {
+            outEncoding.pointee = UInt(kCFStringEncodingInvalidId)
+        }
+        
+        bytes.pointee = result.bytes
+        length.pointee = result.length
+        
+        switch result.deallocator {
+        case .unmap:
+            didMap.pointee = ObjCBool(true)
+        default:
+            didMap.pointee = ObjCBool(false)
+        }
+    }
+}
+#endif
+
+internal func readDataFromFile(path inPath: PathOrURL, reportProgress: Bool, maxLength: Int? = nil, options: Data.ReadingOptions = []) throws -> Data {
+    var attributes: [String : Data] = [:]
+    return try readDataFromFile(path: inPath, reportProgress: reportProgress, maxLength: maxLength, options: options, attributesToRead: [], attributes: &attributes)
+}
+
+internal func readDataFromFile(path inPath: PathOrURL, reportProgress: Bool, maxLength: Int? = nil, options: Data.ReadingOptions = [], attributesToRead: [String], attributes: inout [String: Data]) throws -> Data {
+    let result = try readBytesFromFile(path: inPath, reportProgress: reportProgress, maxLength: maxLength, options: options, attributesToRead: attributesToRead, attributes: &attributes)
+    
+    if result.length == 0 {
+        return Data()
+    } else {
+        return Data(bytesNoCopy: result.bytes!, count: result.length, deallocator: result.deallocator!)
+    }
+}
+
+struct ReadBytesResult {
+    /// Pointer to the read bytes.
+    var bytes: UnsafeMutableRawPointer?
+    
+    /// Number of bytes.
+    /// Matches `Data`'s count type.
+    var length: Int
+    
+    /// The deallocator to use for these bytes, or nil if no deallocator is needed.
+    var deallocator: Data.Deallocator?
+}
+
+internal func readBytesFromFile(path inPath: PathOrURL, reportProgress: Bool, maxLength: Int?, options: Data.ReadingOptions, attributesToRead: [String], attributes: inout [String: Data]) throws -> ReadBytesResult {
+    if inPath.isEmpty {
+        // For compatibility, throw a different error than the perhaps-expected 'file not found' here (41646641)
+        throw CocoaError(.fileReadInvalidFileName)
+    }
+    
+    let fd = try inPath.withFileSystemRepresentation { inPathFileSystemRep in
```

Comment:
I wonder if `withFileSystemRepresentation` is necessary here and elsewhere, even if it only has an effect on Darwin platforms.

I actually think that method should be deprecated. Fundamentally, Foundation should not be trying to second-guess how the filesystem is going to transform a file name. Even the operating system doesn't really know what kind of transformations are going to happen, because that's handled by the (possibly third-party) filesystem driver. You just give it a file name, and if it needs to be transformed, the driver will do it.

AFAICT, other libraries which wrap the POSIX `open` function do not perform this transformation on Darwin platforms (e.g. NodeJS doesn't), so why is it necessary for Foundation to do it?

At the very least, I think this needs documenting, but as explained I think we should investigate avoiding this work altogether.

Reply:
I added `withFileSystemRepresentation` in a prior PR in order to mirror the functionality of the existing APIs that Foundation provides on `String`/`NSString` (but with a swifty interface). I think that we should use `withFileSystemRepresentation` here and other places when calling down to C APIs that accept paths for a few reasons. First, it maintains compatibility with existing Foundation implementations. The existing Foundation framework performs this transformation and changing the implementation as we port it to swift could lead to compatibility-breaking behavior, so a safer bet is to maintain the compatibility here.

But perhaps more importantly, I think it's still the correct thing to do. You are right that the transformations can be handled by the file system driver and that different file systems can behave differently within the same OS, so this implementation isn't _completely_ accurate. That being said, I think this is still the best choice for Darwin. While APFS handles unicode normalization, HFS+ (which is also commonly used on Darwin platforms) does not and failing to normalize strings here (using this special character table) can result in incorrect paths being used. If we don't normalize HFS+ paths, we might not find the file we're looking for due to a mismatch which could be surprising to some developers. In the worst case, we perform a normalization that something like APFS would have done for us, but in the best case we ensure that we normalize for file systems where that is a requirement.

Another benefit is that `withFileSystemRepresentation` is portable - for example, my understanding is that Windows should use a UTF-16 representation and this allows us to have one entry point for conversion to file system representations that can vary based on the platform.

I'm happy to add some documentation to that function as a followup to explain this for maintainers since I can see how that history/context might not be clear from the API itself.

Reply:
Despite the surprising behavior that was my understanding of how it worked. I had some older documentation I was looking to explain this but sadly I can't seem to find that document, and writing a sample app operating on an HFS volume actually seems to prove (from what I can tell) that the HFS+ driver does automatically decompose when creating/reading from the file. So after taking a further look, it does seem that perhaps this normalization might not have been originally necessary (or perhaps this is an artifact of a time when HFS didn't perform that normalization for us). I'll see if I can gather some more info there with examples to thoroughly document it.

That being said - although it looks like I was incorrect about the necessity for interaction with HFS, the compatibility justification still stands for file systems that are normalization-preserving like APFS. If an app exists today that stores a file name and reads/writes to that file using Foundation APIs, the filename will always be the fully decomposed format. Changing this implementation to avoid the decomposition would be a behavioral difference on an APFS volume where normalization is preserved, and I think we'd likely see fallout from changing Foundations' behavior here despite it being different behavior than other libraries that might call `open without this normalization.

I'll do some more investigation on the necessity of this (in case other clients calling `open` might be incorrect in not localizing) but I think that can be handled separately since I think we need to land this `Data` implementation as-is.

Reply:
Thanks for investigating. I hope that this Foundation reimplementation effort can provide an opportunity to reexamine some of these behaviours. I appreciate that this may take time, but I still like to raise the issue so people know about it and look in to it.

I'll just give a quick response to this point:

> That being said - although it looks like I was incorrect about the necessity for interaction with HFS, the compatibility justification still stands for file systems that are normalization-preserving like APFS. If an app exists today that stores a file name and reads/writes to that file using Foundation APIs, the filename will always be the fully decomposed format. Changing this implementation to avoid the decomposition would be a behavioral difference on an APFS volume where normalization is preserved, and I think we'd likely see fallout from changing Foundations' behavior here despite it being different behavior than other libraries that might call `open without this normalization.

While APFS is normalisation-preserving, I believe it is still normalisation-aware. So if you save a file `caf\u{00E9}` (NFC), you can open it using `cafe\u{0301}` (NFD), and it won't allow both versions to exist in the same folder.

The only visible difference between HFS+ and APFS in this matter is if you create a file, then iterate the contents of its containing directory and look for the file using a Unicode-unaware search such as `memcmp`. HFS+ will have normalised the filename, so you'll need to look for the "file system representation" in order to get a match, while APFS will have stored the filename as you gave it.

The problem is that this whole process is ill-advised. Even if you can open a file using the name "X", that doesn't necessarily mean you will find a file literally named "X" if you iterate the directory's contents. The filesystem might transform the filename in any number of ways we can't predict (e.g. a case-insensitive filesystem might normalise everything to uppercase, and some filesystems need to truncate filenames). Users shouldn't be looking up files in that way.

That's why I argue that `withFileSystemRepresentation` should be deprecated. We should put an end to this idea that a directory will contain a particular byte-string, because it isn't true. Once applications no longer operate with that assumption, we wouldn't need to do this normalisation here, even on APFS -- we could respect the APFS designers intent to have their filesystem be normalisation-preserving.

Anyway, that's enough for now. I'd be very interested to know what your investigations uncover; please do @ me here or on the Swift forums if possible.